core/network/src/p2p/gossip/gossip.rs

```diff
@@ -0,0 +1,321 @@
+use crate::p2p::client::{AppResponseCallback, Client};
+use crate::p2p::gossip::{Gossipable, Set};
+use crate::p2p::sdk::{PullGossipRequest, PullGossipResponse};
+use avalanche_types::ids::Id;
+use log::{debug, error};
+use prost::Message;
+use std::error::Error;
+use std::hash::Hash;
+use std::ops::Deref;
+use tokio::sync::Mutex;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::select;
+use tokio::sync::mpsc::{channel, Receiver};
+use tokio::time::interval;
+
+pub struct Config {
+    pub namespace: String,
+    pub frequency: Duration,
+    pub poll_size: usize,
+}
+
+pub struct Gossiper<S: Set + 'static> {
+    config: Config,
+    set: Arc<Mutex<S>>,
+    client: Arc<Mutex<dyn Client>>,
+    stop_rx: Receiver<()>,
+}
+
+unsafe impl<S: Set> Sync for Gossiper<S> {}
+
+unsafe impl<S: Set> Send for Gossiper<S> {}
+
+impl<S> Gossiper<S>
+    where
+        S: Set,
+        S::Item: Default
+{
+    pub fn new(
+        config: Config,
+        set: Arc<Mutex<S>>, // Mutex or RWLock here ?
+        client: Arc<Mutex<dyn Client>>,
+        stop_rx: Receiver<()>,
+    ) -> Self {
+        Self {
+            config,
+            set,
+            client,
+            stop_rx,
+        }
+    }
+
+    pub async fn gossip(&mut self) {
+        let mut gossip_ticker = interval(self.config.frequency);
+
+        loop {
+            select! {
+                _ = gossip_ticker.tick() => {
+                    if let Err(e) = self.execute().await {
+                        error!("Failed to Gossip : {:?}", e);
+                        //ToDo
+
+                    }
+                }
+                _ = self.stop_rx.recv() => {
+                    debug!("Shutting down gossip");
+                    break;
+                }
+            }
+        }
+    }
+
+    async fn execute(&mut self) -> Result<(), Box<dyn Error>> {
+        //ToDo Dummy vec<u8> for now.
+        let bloom = Vec::new();
+
+        let mut request = PullGossipRequest {
+            filter: bloom,
+            salt: Id::default().to_vec(), //ToDo Use default for now
+        };
+
+        let mut msg_bytes = vec![];
+
+        // We scope the lock here to avoid issue later on
+        {
+            let set_guard = self.set.lock().await;
+
+            let elem = set_guard.fetch_elements();
+
+            let filter = elem.serialize().expect("Issue serializing elem");
+            request.filter = filter;
+            request.encode(&mut msg_bytes)?;
+        }
+
+        for _ in 0..self.config.poll_size {
+            {
+                let set = Arc::clone(&self.set.clone());
+
+                // Initialize the callback that will be used upon receiving a response from our gossip attempt
+                let on_response: AppResponseCallback = Arc::new({
```

Comment:
You don't need the extra curly braces here. 

```suggestion
                let on_response: AppResponseCallback = Arc::new(
```

Reply:
Also, you shouldn't have to wrap `Fn` in an `Arc`...